File: 3428-maximum-and-minimum-sums-of-at-most-size-k-subsequences/3428-maximum-and-minimum-sums-of-at-most-size-k-subsequences.cpp
class Solution {
public:
    #define ll long long
    #define all(x) x.begin(), x.end()
    #define vll vector<ll>
    ll mod = 1e9 + 7;

    ll modpow(ll b, ll e) {
        ll ans = 1;
        while (e) {
            if (e & 1) {
                ans = (ans * b) % mod;
            }
            b = (b * b) % mod;
            e >>= 1;
        }
        return ans;
    }
    
    int minMaxSums(vector<int>& nums, int k) {
        sort(all(nums));
        int n = nums.size();
        vll fact(n+1, 1), invFact(n+1, 1);
        for (ll i = 1; i <= n; i++) {
            fact[i] = (fact[i-1] * i) % mod;
        }
        invFact[n] = modpow(fact[n], mod - 2);
        for (ll i = n - 1; i >= 0; i--) {
            invFact[i] = (invFact[i+1] * (i+1)) % mod;
        }
        function<ll(ll,ll)> counter = [&](ll n, ll r) {
            if (r > n || r < 0) return 0ll;
            return (ll)(fact[n] * invFact[r] % mod) * invFact[n-r] % mod;
        };
        ll ans = 0;
        for (ll i = 0; i < n; i++) {
            for (ll j = 1; j <= k; j++) {
                ll cnt = counter(n-i-1, j-1);
                ans = (ans + nums[i] * cnt) % mod;
            }
            for (ll j = 1; j <= k; j++) {
                ll cnt = counter(i, j-1);
                ans = (ans + nums[i] * cnt) % mod;
            }
        }
        return ans;
    }
};

File: 3428-maximum-and-minimum-sums-of-at-most-size-k-subsequences/3428-maximum-and-minimum-sums-of-at-most-size-k-subsequences.py
MOD = 10**9 + 7

def modpow(b, e):
    ans = 1
    while e > 0:
        if e & 1:
            ans = (ans * b) % MOD
        b = (b * b) % MOD
        e >>= 1
    return ans

class Solution:
    def minMaxSums(self, nums: List[int], k: int) -> int:
        nums.sort()
        n = len(nums)
        
        fact = [1] * (n + 1)
        invFact = [1] * (n + 1)
        
        for i in range(1, n + 1):
            fact[i] = (fact[i - 1] * i) % MOD
        
        invFact[n] = modpow(fact[n], MOD - 2)
        for i in range(n - 1, -1, -1):
            invFact[i] = (invFact[i + 1] * (i + 1)) % MOD

        def counter(n, r):
            if r > n or r < 0:
                return 0
            return (fact[n] * invFact[r] % MOD) * invFact[n - r] % MOD

        ans = 0

        for i in range(n):
            for j in range(1, k + 1):
                cnt = counter(n - i - 1, j - 1)
                ans = (ans + nums[i] * cnt) % MOD
            for j in range(1, k + 1):
                cnt = counter(i, j - 1)
                ans = (ans + nums[i] * cnt) % MOD

        return ans